stl/inc/algorithm

```diff
@@ -961,6 +941,340 @@ namespace ranges {
+            _Rng1&& _Range1, _Rng2&& _Range2, _Pr _Pred = {}, _Pj1 _Proj1 = {}, _Pj2 _Proj2 = {}) const {
+            if constexpr (sized_range<_Rng1> && sized_range<_Rng2>) {
+                const auto _Count = _RANGES distance(_Range1);
+                if (_RANGES distance(_Range2) != _Count) {
+                    return false;
+                }
+
+                return _Sized_ranges(_Get_unwrapped(_RANGES begin(_Range1)), _Get_unwrapped(_RANGES end(_Range1)),
+                    _Get_unwrapped(_RANGES begin(_Range2)), _Get_unwrapped(_RANGES end(_Range2)), _Count,
+                    _Pass_fn(_Pred), _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            } else {
+                return _Unsized_ranges(_Get_unwrapped(_RANGES begin(_Range1)), _Get_unwrapped(_RANGES end(_Range1)),
+                    _Get_unwrapped(_RANGES begin(_Range2)), _Get_unwrapped(_RANGES end(_Range2)), _Pass_fn(_Pred),
+                    _Pass_fn(_Proj1), _Pass_fn(_Proj2));
+            }
+        }
+
+    private:
+        template <class _It1, class _Se1, class _It2, class _Se2, class _Pr, class _Pj1, class _Pj2>
+        _NODISCARD static constexpr bool _Sized_ranges(_It1 _First1, _Se1 _End1, _It2 _First2, _Se2 _End2,
+            iter_difference_t<_It1> _Count, _Pr _Pred, _Pj1 _Proj1, _Pj2 _Proj2) {
+            _STL_INTERNAL_CHECK(_RANGES distance(_First1, _End1) == _Count);
+            _STL_INTERNAL_CHECK(_RANGES distance(_First2, _End2) == _Count);
+
+            for (;;) { // Trim matching prefixes
+                if (_Count == 0) { // Everything matched
+                    return true;
+                }
+
+                if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_First1), _STD invoke(_Proj2, *_First2))) { // mismatch
+                    if (_Count == 1) { // Single non-matching characters remain; not a permutation
+                        return false;
+                    }
+
+                    break;
+                }
+
+                ++_First1;
+                ++_First2;
+                --_Count;
+            }
+            // If we get here, _Count > 1 and initial elements do not match.
+
+            if constexpr (bidirectional_iterator<_It1> && bidirectional_iterator<_It2>) {
+                // Determine final iterator values
+                auto _Last1 = _Find_last_iterator(_First1, _End1, _Count);
+                auto _Last2 = _Find_last_iterator(_First2, _End2, _Count);
+
+                for (;;) { // Trim matching suffixes
+                    --_Last1;
+                    --_Last2;
+                    if (!_STD invoke(_Pred, _STD invoke(_Proj1, *_Last1), _STD invoke(_Proj2, *_Last2))) { // mismatch
+                        ++_Last1;
+                        ++_Last2;
+                        break;
+                    }
+
+                    if (--_Count == 1) {
+                        return false; // initial elements did not match when trimming prefixes
+                    }

```

Comment:
I find that much more complex than the prefix loop. 

I guess the difference is an superfluous `if (_Count == 1)` check at the beginning

Reply:
They're nearly mirrors, which is interestingly indicative of doing the same thing but backwards. The prefix loop is "check if we've run out of data, compare the current elements, bump current iterators and reduce count" whereas this one is "bump current iterators, compare the current elements, check if we've run out of data." Notably the suffix loop "runs out of data" when `_Count` is one instead of zero since the prefix loop already proved that the initial elements do not match.

Do you think it would help if we put the increments of `_Last1` and `_Last2` after the loop body instead of before the `break`?

Reply:
I honestly did not see the mirroring. That said this is the least complex part of the algorithm

Reply:
Is there a specific change that you think would help, or are you happy enough with this as-as?

Reply:
Not really, i find your mirroring argument convincing